### PR TITLE
[Videoversion] Fixed artwork issue in VideoVersion dialog

### LINKED ISSUE #24136: [VideoVersion] "Resume"  from "Dialog Video Info" start always from beginning with multi versions
### Describe the bug
Here is a clear and concise description of what the problem is:

If you you want resume a movie that has more than 1 version from "Dialog Video Info" the movie start alway's from the beginning.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I expect that the movie resume from the last position. 
This works fine from the widget at home from "in progess movies" widget

## Actual Behavior

"Resume"  from "Dialog Video Info" start always from begining with multi version set 

### To Reproduce

1.  have a movie with 2 versions 
2. start to read one of this version
3. stop reading
4. open "information" for this movie"
5. click on "play" button
6. "Resume from" / "play from beginning" is displayed
7. Select "Resume from"
8. The "DialogVideoVersion" dialog box opens automatically
9. Then select one or the other version launch the movie at the beginning

### Debuglog
The debuglog can be found here:

https://paste.kodi.tv/bayusomipe.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

It's a fresh feature. So no urgent here. This issue is FYI only. 
Thanks for all works  :)

### Your Environment
Used Operating system:

 - [X] Windows

 - Operating system version/name:
 - Kodi version: 21.0-BETA1 (20.90.801) Git:20231122-8f075c5358)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by xodidox**
## Description

[VideoVersion] Fixed artwork issue in VideoVersion dialog

Currently when you changed artwork in the dialog, the updated artwork is not available unless close/reopen the dialog; in other cases (rename, for example), the artwork will disappear.
## Motivation and context

Feedback
## How has this been tested?

Runtime tested on MacBook
## What is the effect on users?

Improved user experience
## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/video/VideoDatabase.cpp
Comment (enen92) [inline xbmc/video/VideoDatabase.cpp:11902]: Maybe use std::optional, better than returning -1

Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:11902]: The id value -1 is used everywhere for VideoVersion to represent the invalid id value.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp:743]: Although this might work (did no runtime test) I consider this just another workaround for the actual problem that the video version dialog needs to be triggered earlier, by the implementation of the videoplayactionprosessor (base class).

Comment (ksooo) [reply xbmc/video/dialogs/GUIDialogVideoInfo.cpp:743]: I feel this is a bandaid for only one place out of many where we also see this problem.

**Comment by ksooo** [comment]
@xodidox may I suggest that you remove the fix for the resume bug and let me take over? I have an idea how to fix it properly and for all places it can occur but this is more than just a one-liner and I have no time to teach you. I tried explaining multiple times, but it seems I'm not getting the message across, which is okay as it is not something trivial. I think we will be faster if I come up with the implementation and you can have a look what I ment all the time when I was saying "this belongs into the videoplayactionprocessor." if you are interested to learn.

**Comment by ksooo** [comment]
@xodidox I had a closer look at how playback currently works for video versions and came to the conclusion that making it compatible with general play and select action concept will be a larger undertaking (currently it is really very special handling, unfortunately, not fitting well into the Video*Actionprocessor design)  and I have no time for this currently. So, can you please open another PR with the fix for the resume issue (please, di not append to this PR!) so we have it at least consistent with how you implemented it as of now.

**Comment by jjd-uk** [comment]
Can you also given a brief description of what the artwork problem is that this is fixing.

**Comment by xodidox** [comment]
@jjd-uk updated the description.

**Comment by jjd-uk** [comment]
Excellent, thanks for the description. I think I may have seen that happen, and it was on my todo list to see if I could replicate in a consistent manner.

**Comment by xodidox** [comment]
@ksooo @enen92 @jjd-uk is this good to merge?

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp:11849]: This change seems unrelated to the fix. Please remove from PR or put into separate commit with a speaking commit message.

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp:11912]: I think the whole method can be implemented much more simple using `CDatabase::GetSingleValueInt`.

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp:11937]: `const int`

error handling?

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp:11949]: `std::find_if`

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoVersion.cpp:595]: `std::find_if`

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoVersion.cpp:603]: `std::find_if`

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoVersion.cpp:605]: `return {};`

File: xbmc/video/dialogs/GUIDialogVideoVersion.h
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoVersion.h:49]: Could this method made be `const`?

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:11914]: CDatabase::GetSingleValueInt returns 0 when failed, we need -1.

**Review by jjd-uk — APPROVED** [review]
Can't say anything about the C++ side of things but from a user perspective this does seemed to have improved artwork handling. Adding versions of movies I remembering having problems with artwork no longer seems to have any problems.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11914]: I don't think it's viable to keep adding a new function and SQL query every time we need something slightly different.
The movie table already has helper functions. `GetDetailsByTypeAndId()` should be able to do the job?

Right now `GetDetailsByTypeAndId()` doesn't have a parameter for the bit field getDetails to avoid unnecessary additional queries, but the functions it calls already accept it (see `GetMovieInfo()` etc.) Please add that parameter, and this function can send a 0 as bit field value because all it needs is in the movie table. That part should be 1 dedicated commit that comes before the commit that contains your fix.

**Comment by xodidox** [comment]
@ksooo when can I expect this to be merged?

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [inline xbmc/video/VideoDatabase.cpp:11911]: just a cosmetic, but could we invert the condition here? So instead of
```
if (itemType == VideoDbContentType::MOVIES) {
   // some logic
} else {
  return -1;
}
```
better use this logic
```
if (itemType != VideoDbContentType::MOVIES)
   return -1;

 // some logic
```

the current code is not deeply nested, but it's always better to avoid unnecessary code nesting if possible.
Just noticed that you could even move this condition also into the previous ` if (!m_pDB || !m_pDS)` condition and clean up the code even more.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:11938]: I could have missed something but looks to me like GetDetailsByTypeAndId() already does what this revised function is doing (return FileItem for a given item type and id)

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:11911]: This is for future extension (TV Shows/Seasons extras, Episode versions/extras, etc.) for Video Versions feature. All implementation related with this future extension is implemented in this way.

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:11938]: This one just picks the default item in result of GetVideoVersions. The GetVideoVersions has special handling for Video Versions file items. It's not in GetDetailsByTypeAndId.

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:11952]: Fixed.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11939]: OK I see that it makes up the information of the VideoInfoTags and it looks like you're trying to consolidate with other existing code. Looking forward to the next steps.

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:11911]: I'd say keep it clean for now - nobody knows how the data structure will look like in future. But that's just my take on it

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (rmrector) [inline xbmc/video/dialogs/GUIDialogVideoVersion.cpp:287]: `CVideoThumbLoader` is an implementation of `CBackgroundInfoLoader`, the `void Load(CFileItemList& items)` function will populate images in a background thread. Keep it as a member variable of this class and make sure the thread is stopped when the window closes or another list is loaded with the below snippet.
```cpp
if (m_thumbLoader.IsLoading())
  m_thumbLoader.StopThread();
```

See how [CGUIWindowVideoBase](https://github.com/xbmc/xbmc/blob/master/xbmc/video/windows/GUIWindowVideoBase.cpp) uses `m_thumbLoader`, you probably don't need the `SetObserver` part. I'm not sure how easy it is to combine both lists you have here, but that will help loading the whole list in `ConvertVideoVersion`.

Comment (xodidox) [reply xbmc/video/dialogs/GUIDialogVideoVersion.cpp:287]: Thanks @rmrector for this information. The artwork issue here is for video version dialog, there is not perf issue here as the items count is very limited. It is the original design issue with a unique_ptr for video version dialog, which cause issue when change the item (as reported in the related issues).

For ConvertVideoVersion artwork issue, with the feedback from forum, even if we fixed the artwork issue, trying to find a movie in a list of at least several hundred (or several thousand as reported in forum) has a usability issue. I'll create a new PR to fix that after this PR merged by only providing the detected same movies in the select dialog, instead of list all movies. User can always add a new version via "Add version" button in video version dialog to find the video file directly.

**Comment by ksooo** [comment]
Looking at this PR, I’m honestly asking myself why we need a shitload of new video database functionality to fix a simple refresh of data. Not saying this is wrong, but before this gets merged somebody should convince me that there is no better (less invasive) way of fixing the issue of artwork refresh.

**Comment by ksooo** [comment]
@xodidox, maybe I misunderstood the issue, but here is how I would fix it. No new database stuff needed, much less invasive than you approach. What do you think?

```diff
diff --git a/xbmc/video/dialogs/GUIDialogVideoVersion.cpp b/xbmc/video/dialogs/GUIDialogVideoVersion.cpp
index e9c0f7f781..19db2de837 100644
--- a/xbmc/video/dialogs/GUIDialogVideoVersion.cpp
+++ b/xbmc/video/dialogs/GUIDialogVideoVersion.cpp
@@ -37,7 +37,6 @@
 #include "utils/Variant.h"
 #include "utils/log.h"
 #include "video/VideoInfoTag.h"
-#include "video/VideoThumbLoader.h"
 #include "video/VideoUtils.h"
 #include "video/dialogs/GUIDialogVideoInfo.h"
 #include "video/guilib/VideoPlayActionProcessor.h"
@@ -79,6 +78,12 @@ bool CGUIDialogVideoVersion::OnMessage(CGUIMessage& message)
   {
     case GUI_MSG_WINDOW_DEINIT:
     {
+      if (m_thumbLoaderVersions.IsLoading())
+        m_thumbLoaderVersions.StopThread();
+
+      if (m_thumbLoaderExtras.IsLoading())
+        m_thumbLoaderExtras.StopThread();
+
       ClearVideoVersionList();
       break;
     }
@@ -223,6 +228,12 @@ void CGUIDialogVideoVersion::RefreshVideoVersionList()
   // clear current version list
   ClearVideoVersionList();
 
+  if (m_thumbLoaderVersions.IsLoading())
+    m_thumbLoaderVersions.StopThread();
+
+  if (m_thumbLoaderExtras.IsLoading())
+    m_thumbLoaderExtras.StopThread();
+
   const int dbId = m_videoItem->GetVideoInfoTag()->m_iDbId;
   MediaType mediaType = m_videoItem->GetVideoInfoTag()->m_type;
   VideoDbContentType itemType = m_videoItem->GetVideoContentType();
@@ -232,16 +243,20 @@ void CGUIDialogVideoVersion::RefreshVideoVersionList()
                               VideoVersionItemType::PRIMARY);
   m_primaryVideoVersionList->SetContent(CMediaTypes::ToPlural(mediaType));
 
-  // refresh primary version list
-  CGUIMessage msg1(GUI_MSG_LABEL_BIND, GetID(), CONTROL_LIST_PRIMARY_VERSION, 0, 0,
-                   m_primaryVideoVersionList.get());
-  OnMessage(msg1);
-
   // get extras version list
   m_database.GetVideoVersions(itemType, dbId, *m_extrasVideoVersionList,
                               VideoVersionItemType::EXTRAS);
   m_extrasVideoVersionList->SetContent(CMediaTypes::ToPlural(mediaType));
 
+  // Reload artwork async
+  m_thumbLoaderVersions.Load(*m_primaryVideoVersionList);
+  m_thumbLoaderExtras.Load(*m_primaryVideoVersionList);
+
+  // refresh primary version list
+  CGUIMessage msg1(GUI_MSG_LABEL_BIND, GetID(), CONTROL_LIST_PRIMARY_VERSION, 0, 0,
+                   m_primaryVideoVersionList.get());
+  OnMessage(msg1);
+
   // refresh extras version list
   CGUIMessage msg2(GUI_MSG_LABEL_BIND, GetID(), CONTROL_LIST_EXTRAS_VERSION, 0, 0,
                    m_extrasVideoVersionList.get());
@@ -440,17 +455,7 @@ void CGUIDialogVideoVersion::ChooseArt()
           std::make_shared<CFileItem>(*m_selectedVideoVersion)))
     return;
 
-  // update the thumbnail
-  CGUIControl* control = GetControl(CONTROL_IMAGE_THUMB);
-  if (control)
-  {
-    CGUIImage* imageControl = static_cast<CGUIImage*>(control);
-    imageControl->FreeResources();
-    imageControl->SetFileName(m_selectedVideoVersion->GetArt("thumb"));
-  }
-
-  CGUIMessage msg(GUI_MSG_NOTIFY_ALL, 0, 0, GUI_MSG_REFRESH_THUMBS);
-  CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+  RefreshVideoVersionList();
 }
 
 void CGUIDialogVideoVersion::AddVersion()
diff --git a/xbmc/video/dialogs/GUIDialogVideoVersion.h b/xbmc/video/dialogs/GUIDialogVideoVersion.h
index 27e38aae79..058f5599d7 100644
--- a/xbmc/video/dialogs/GUIDialogVideoVersion.h
+++ b/xbmc/video/dialogs/GUIDialogVideoVersion.h
@@ -12,6 +12,7 @@
 #include "guilib/GUIDialog.h"
 #include "media/MediaType.h"
 #include "video/VideoDatabase.h"
+#include "video/VideoThumbLoader.h"
 
 #include <memory>
 
@@ -63,4 +64,6 @@ private:
   std::unique_ptr<CFileItemList> m_extrasVideoVersionList;
   std::unique_ptr<CFileItem> m_defaultVideoVersion;
   std::shared_ptr<CFileItem> m_selectedVideoVersion;
+  CVideoThumbLoader m_thumbLoaderVersions;
+  CVideoThumbLoader m_thumbLoaderExtras;
 };
```

**Comment by xodidox** [comment]
Rebased

@ksooo The CVideoDatabase change just removed the duplicated code and add a new function to get the default version id. The artwork issue as described in the PR, it's not the perf issue as the items are very limited for the video version dialog, thus no perf issue. With this PR, the versions information will be only retrieved once from DB, and all other queries are done in the retrieved information in CGUIDialogVideoVersion, thus there is no inconsistence issue as current design. This also requires the change of unique_ptr to shared_ptr.

**Comment by CrystalP** [comment]
@xodidox don't be discouraged, this PR changes multiple things at the same time, which makes it more difficult to understand and review.

Please try structuring with one commit per logical change, with descriptive commit message. There is no absolute rule how to do it, but here: refactor = 1 commit, fix for artwork = 1 commit should be good (maybe not in that sequence though, depends if the refactor is a cleanup before the change, or meant to cleanup after the functionality change).
Depending on how it ties with the rest, adding the new function for default version id: maybe it's own commit.

**Comment by ksooo** [comment]
@xodidox I carefully studied the changes you did in this PR. 

The database changes are not needed to fix the problem. But the removal of duplicate code there introduces a performance decrease when retrieving the default version. Before the PR, you made exactly one query that precisely picked the one record you needed. Now, you obtain all versions from the db and after that you use std::find_if to find  the one record you need. Why should we want this?

Regarding the thumb loader. Although there is currently no performance problem, why not improve the code, especially as the person with the best knowledge in that area suggested to change this?

All in all I find the changes you made not necessary, making the code more complicated and harder to understand. 

At the end it is your decision, but please don’t expect my approval for this PR in this state.

**Comment by ksooo** [comment]
> At the end it is your decision, but please don’t expect my approval for this PR in this state.

To be clear, I'm not blocking merge of this PR. I just said that I would have done this different. If some body else with merge rights wants to hit the green button, also fine for me.

**Comment by ksooo** [comment]
@xodidox it's a pity. As I said: "I find some of the changes you made not necessary, making the code more complicated and harder to understand" - you see the word **some** here?

Now somebody else needs to come up with a PR for the parts not in question here.

**Pull request closed without merging**
